### PR TITLE
Pass tune-cpu to LLVM

### PR DESCRIPTION
**Pull request by Artoria2e5**
I think this is how it should work...

See https://internals.rust-lang.org/t/expose-tune-cpu-from-llvm/13088 for the background. Or the documentation diff.

### CODE REVIEW

**Comment by lcnr** [comment]
Not an area I am familiar with, maybe r? @nagisa

File: compiler/rustc_session/src/options.rs [hunk 1]
```diff
@@ -810,6 +810,8 @@ options! {CodegenOptions, CodegenSetter, basic_codegen_options,
     target_feature: String = (String::new(), parse_target_feature, [TRACKED],
         "target specific attributes. (`rustc --print target-features` for details). \
         This feature is unsafe."),
+    tune_cpu: Option<String> = (None, parse_opt_string, [TRACKED],
+        "select processor to schedule for (`rustc --print target-cpus` for details)"),
```
Comment (jonas-schievink) [inline compiler/rustc_session/src/options.rs:814]: This should start out as a debugging option, otherwise it would be instantly stable

Comment (Artoria2e5) [reply compiler/rustc_session/src/options.rs:814]: Makes sense. Let me see how that works...

File: src/doc/rustc/src/codegen-options/index.md [hunk 2]
```diff
@@ -530,6 +530,19 @@ This also supports the feature `+crt-static` and `-crt-static` to control
 Each target and [`target-cpu`](#target-cpu) has a default set of enabled
 features.
 
+## tune-cpu
```
Comment (nagisa) [inline src/doc/rustc/src/codegen-options/index.md:533]: Moving this option to `-Z` made it unstable and not-actually-a-codegen-option. At the very least this should be annotated as unstable.

File: compiler/rustc_codegen_llvm/src/attributes.rs [hunk 3]
```diff
@@ -194,6 +194,16 @@ pub fn apply_target_cpu_attr(cx: &CodegenCx<'ll, '_>, llfn: &'ll Value) {
     );
 }
 
+pub fn apply_tune_cpu_attr(cx: &CodegenCx<'ll, '_>, llfn: &'ll Value) {
```
Comment (nagisa) [inline compiler/rustc_codegen_llvm/src/attributes.rs:197]: This applies the attribute regardless of whether the option is specified or not. Is there a good reason why we're applying it even if the `-Ztune-cpu` is not specified?

**Review by nagisa — COMMENTED** [review]
This needs a dedicated test which uses the newly added `-Z` flag.

File: compiler/rustc_codegen_llvm/src/attributes.rs (hunk 3 above)
Comment (Artoria2e5) [reply compiler/rustc_codegen_llvm/src/attributes.rs:197]: LLVM should handle the fallback value, so no real reason for that except testability. I guess I will make a separate thing for it.

File: src/doc/rustc/src/codegen-options/index.md (hunk 2 above)
Comment (Artoria2e5) [reply src/doc/rustc/src/codegen-options/index.md:533]: Annotated, but not moved yet since I don't know where unstable opts should be documented.

**Comment by nagisa** [comment]
This overall LGTM. There are remaining compilation issues remaining before this can land and the `tune-cpu` documentation still needs a note that it is unstable. Please also squash.

File: src/doc/rustc/src/codegen-options/index.md [hunk 4]
```diff
@@ -530,6 +530,21 @@ This also supports the feature `+crt-static` and `-crt-static` to control
...
+
+This instructs `rustc` to schedule code specifically for a particular processor.
+This does not affect what instruction sets are used or the compatibility, but
+might make your code run a bit faster on the selected CPU.
+
+You can run `rustc --print target-cpus` to see the valid options to pass
+here. Additionally, `native` can be passed to use the processor of the host
```
Comment (Artoria2e5) [inline src/doc/rustc/src/codegen-options/index.md:540]: Might also be a good idea to explain `generic` in here and `target-cpu`: it's basically a profile for ancient instructions tuned for modern CPU.

File: compiler/rustc_codegen_llvm/src/attributes.rs [hunk 5]
```diff
@@ -300,6 +312,9 @@ pub fn from_fn_attrs(cx: &CodegenCx<'ll, 'tcx>, llfn: &'ll Value, instance: ty::
     // Without this, ThinLTO won't inline Rust functions into Clang generated
     // functions (because Clang annotates functions this way too).
     apply_target_cpu_attr(cx, llfn);
+    // tune-cpu is only conveyed through the attribute for our purpose.
+    // The target doesn't care; the subtarget reads our attribute.
+    apply_tune_cpu_attr(cx, llfn);
```
Comment (cuviper) [inline compiler/rustc_codegen_llvm/src/attributes.rs:317]: Does tune-cpu have any of the same issues as noted for target-cpu above, affecting cross-language LTO?

Comment (Artoria2e5) [reply compiler/rustc_codegen_llvm/src/attributes.rs:317]: Since it's the only way this gets conveyed in Clang, kind of?

Comment (cuviper) [reply compiler/rustc_codegen_llvm/src/attributes.rs:317]: More directly, does the mismatched presence or absence of this attribute interfere with optimizations like inlining?

Comment (nagisa) [reply compiler/rustc_codegen_llvm/src/attributes.rs:317]: No.

**Comment by nagisa** [comment]
Still blocked on a squash.

**Comment by Artoria2e5** [comment]
Grrr https://github.community/t/set-receive-shallowupdate-on-to-allow-pushing-from-shallow-clones/2424/2

**Comment by Artoria2e5** [comment]
wha--what?

**Comment by khyperia** [comment]
Wait, can I ask why `tune_cpu` was added to the backend interface? It doesn't seem to be called anywhere (deleting it and running `./x.py check` seems fine). Same with `target_cpu`, actually, it doesn't seem to be called anywhere...

**Comment by Artoria2e5** [comment]
I agree that this should probably be removed. I just added it because `target_cpu` was there, but yeah...

**Pull request merged**